sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/implementation/ConcurrencyLimitingSpliterator.java

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// Code generated by Microsoft (R) AutoRest Code Generator.
+
+package com.azure.monitor.ingestion.implementation;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.Consumer;
+
+public class ConcurrencyLimitingSpliterator<T> implements Spliterator<T> {
+    private List<T> items;
+    private int begin;
+    private final int step;
+    private final ConcurrentLinkedQueue<ConcurrencyLimitingSpliterator> spliterators;
+
+    public ConcurrencyLimitingSpliterator(List<T> items, int concurrency) {
+        Objects.requireNonNull(items, "'items' cannot be null.");
+        if (concurrency == 0) {
+            throw new IllegalArgumentException("'concurrency' must be a positive number.");
+        }
+        this.items = items;
+        this.step = concurrency;
+
+        if (concurrency > 1) {
+            spliterators = new ConcurrentLinkedQueue<>();
+            for (int i = 0; i < Math.min(concurrency, items.size()) - 1; i++) {
+                spliterators.add(new ConcurrencyLimitingSpliterator(items, i, concurrency));
+            }
+        } else {
+            spliterators = null;
+        }
+
+        this.begin = Math.min(concurrency, items.size()) - 1;
+    }
+
+    private ConcurrencyLimitingSpliterator(List<T> items, int begin, int step) {
+        this.items = items;
+        this.begin = begin;
+        this.step = step;
+        this.spliterators = null;
+    }
+
+    @Override
+    public boolean tryAdvance(Consumer action) {
+        if (begin >= items.size()) {
+            return false;
+        }
+
+        action.accept(items.get(begin));
+        begin += step;
+        return true;
+    }
+
+    @Override
+    public Spliterator trySplit() {
+        return spliterators != null ? spliterators.poll() : null;
+    }
+
+    @Override
+    public long estimateSize() {
+        int currentSpliteratorSize = (items.size() - begin) / step + 1;
+        if (spliterators == null) {
+            // inner spliterator
+            return currentSpliteratorSize;
+        }
+
+        // add size of all unsplitted spliterators
+        return currentSpliteratorSize + spliterators.size() * items.size() / step + 1;
+    }
+
+    @Override
+    public int characteristics() {
+        return NONNULL | ORDERED | SIZED | SUBSIZED;
```

Comment:
Does this need to include the `Concurrent` characteristic as well?

Reply:
Concurrent characteristics means that underlying data can be changed concurrently, which is not the case. 
For parallel execution, we need parallel stream, but spliterator does not really care about execution.